Troubleshooting Guide: Versioning (#223)
microsoft/vcpkg-docs: vcpkg/users/versioning-troubleshooting.md

---
title: Troubleshoot versioning and feature selection issues
description: Troubleshooting guide for common versioning issues
author: JavierMatosD
ms.author: javiermat
ms.date: 11/30/23
ms.prod: vcpkg
ms.topic: troubleshooting-general
---

# Versioning troubleshooting guide

> This guide is intended for users experiencing issues with [versioning](./versioning.md).

Reviewer: Maybe we should add a section about inspecting the version file fora port, so users know to look for `versions/<p>-/<port>`

### Cause: Requesting a Non-Existent Version of a Package
<!-- 
Steps to reproduce:
1. Open or create a vcpkg.json file for your project.
2. Specify a dependency with a version that does not exist in the vcpkg repository. For example, set the version of the fmt library to a non-existent version like "version>=": "100.0.0".
3. vcpkg install
-->

When a version specified in the manifest file does not exist in the vcpkg version database, vcpkg will fail to resolve the dependency and throw an error resembling the following:

```console
error: no version database entry for fmt at 100.0.0
Available versions:
  10.1.1
  10.1.0
  10.0.0
  9.1.0#1
  9.1.0
  9.0.0
  8.1.1#2
  8.1.1#1
  ...
See `vcpkg help versioning` or https://learn.microsoft.com/vcpkg/users/versioning for more information.
```

To resolve the issue:

> 1. Check available versions:
>    - This can be done by visiting the vcpkg GitHub repository or by noting the versions listed in the error message provided by vcpkg.
> 2. Update manifest file:
>    - Edit your vcpkg.json file.
>    - Change the specified version to one that is available in the vcpkg repository. For example, change from "version>=": "100.0.0" to "version>=": "10.1.1".
> 3. Run vcpkg install:
>    - Execute the `vcpkg install` command again in your terminal or command prompt.

Reviewer: ```suggestion
1. Update the versions database:
   - The version you want may not be in your local copy of the versions database. In that case, run the `git pull` command to update the [vcpkg registry](<https://github.com/Microsoft/vcpkg>) to the latest commit.
2. Check available versions:
   - Choose one of the versions available in the versions database.
3. Update manifest file:
   - Edit your vcpkg.json file.
   - Change the specified version to one that is available in the vcpkg repository. For example, change from "version>=": "100.0.0" to "version>=": "10.1.1".
4. Run vcpkg install:
   - Execute the `vcpkg install` command again in your terminal or command prompt.
```

Suggest updating the version database via `git pull` as the first step.

## <a name="version-scheme-conflict"></a> Version scheme conflict in dependencies

A version scheme conflict occurs when the version specified in the `vcpkg.json` file for a dependency uses a different versioning scheme than the one used in the vcpkg repository's baseline version. This results in an error as vcpkg cannot compare versions across different schemes.

### Cause: Specifying version constraint across different schemes
<!--
Steps to reproduce:
1. Open or create a vcpkg.json file for your project.
2. Specify boost-regex as a dependency with version>= 1.75.0
3. vcpkg install
-->

> When a version>= constraint is specified in the `vcpkg.json` for a dependency that is useing a different version scheme than the one used in vcpkg's baseline version.

Reviewer: ```suggestion
A declared `version>=` constraint uses a different version scheme than the one used in the baseline version. When this happens, vcpkg is unable to determine which version is greater or equal than the other.
```

---
title: Troubleshoot versioning and feature selection issues
description: Troubleshooting guide for common versioning issues
author: JavierMatosD
ms.author: javiermat
ms.date: 11/30/2023
ms.prod: vcpkg
ms.topic: troubleshooting-general
---

# Versioning troubleshooting guide

This guide is intended for users experiencing issues with [versioning](./versioning.md).

> ## Inspecting the version file for a port

Reviewer: We really need a command that does this for the user, and it needs to work with registries.

Example:
```
./vcpkg x-package-info --versions --x-json fmt
{
  "versions": [
    {
      "git-tree": "dfe9aa860f5a8317f341a21d317be1cf44e89f18",
      "version": "10.1.1",
      "port-version": 0
    },
    {
      "git-tree": "d8209382ca192426861c6aa91d217ea8512bda46",
      "version": "10.1.0",
      "port-version": 0
    },
  //etc
}
```

> ## Inspecting the version file for a port

Reviewer: ```suggestion
## <a name="inspect-versions-file"></a> Inspecting the version file for a port
```

The following process works for vcpkg registry ports. For more detailed information, see our [registry](../maintainers/registries.md#git-registries) documentation. This process involves checking the available versions of a package in the vcpkg repository.

How to inspect:
1. Navigate to the `vcpkg/versions` directory.
2. Locate the port folder:

>    - Within the `versions` directory, ports are organized alphabetically in folders. Find the folder corresponding to the first letter of your desired port. For example, for `fmt`, look in the `f-` folder.

Reviewer: ```suggestion
   - Find the folder corresponding to the first letter of the port. For example, for `fmt` open the folder named `f-`.
```

3. Open the ports version file:

>    - In the appropriate alphabetical folder, locate the JSON file for your port. For `fmt`, this would be `fmt.json.`

Reviewer: ```suggestion
   - Locate the JSON file with the same name of the port. For example, the `fmt` versions file is named `fmt.json.`
```

> What you will find:
> 

Reviewer: ```suggestion
```

> In the port's version file, you'll see a list of available versions with details like version numbers and corresponding commit hashes. This information is crucial for understanding what versions you can specify in your project's `vcpkg.json` file.

Reviewer: ```suggestion
In the port's version file, you'll see a list of available versions with details like version numbers and corresponding [Git tree-object hashes](https://git-scm.com/book/en/v2/Git-Internals-Git-Objects#_tree_objects). This information is crucial for understanding what versions you can specify in your project's `vcpkg.json` file.
```

> In the port's version file, you'll see a list of available versions with details like version numbers and corresponding [Git tree-object hashes](https://git-scm.com/book/en/v2/Git-Internals-Git-Objects#_tree_objects). This information is crucial for understanding what versions you can specify in your project's `vcpkg.json` file.

Reviewer: ```suggestion
The port's version file contains a list of available versions with details like version tags and their corresponding [Git tree-object hash](https://git-scm.com/book/en/v2/Git-Internals-Git-Objects#_tree_objects). This information is required by vcpkg to retrieve specific port versions. Only versions contained in this list are available to use in your manifest files.
```

How to use this information:
- Choose a version from the list that meets your projects needs.
- Reflect this in your projects `vpckg.json` file under the appropriate dependency.

> - After updating the manifest file, run `vcpkg install` to install newly specified version.

Reviewer: Remove this since it makes more sense to have these instructions in the resolution steps of "non-existent version failures".

For more details on versioning, see our reference documentation:
- [versioning concepts](./versioning.concepts.md)
- [versioning](./versioning.md)

For more details on using a manifest, see [manifest](./manifests.md)

## <a name="non-existent-version"></a> Cause: Requesting a non-existent version of a package
<!-- 
Steps to reproduce:
1. Open or create a vcpkg.json file for your project.
2. Specify a dependency with a version that does not exist in the vcpkg repository. For example, set the version of the fmt library to a non-existent version like "version>=": "100.0.0".
3. vcpkg install
-->

> When a version specified in the manifest file does not exist in the vcpkg version database, vcpkg will fail to resolve the dependency and throw an error resembling the following:

Reviewer: ```suggestion
When a version specified in the manifest file does not exist in the vcpkg version database, vcpkg fails to resolve the dependency and produces an error message resembling the following:
```

   - Edit your `vcpkg.json` file.
   - Change the specified version to one that is available in the vcpkg repository. For example, change from "version>=": "100.0.0" to "version>=": "10.1.1".  
4. Run vcpkg install:
   - Execute the `vcpkg install` command again in your terminal or command prompt.  

## <a name="version-scheme-conflict"></a> Cause: Specifying version constraint across different schemes
<!--
Steps to reproduce:
1. Open or create a vcpkg.json file for your project.
2. Specify boost-regex as a dependency with version>= 1.75.0
3. vcpkg install
-->
A version scheme conflict occurs when the version specified in the `vcpkg.json` file for a dependency uses a different versioning scheme than the one used in the vcpkg repository's baseline version. This results in an error as vcpkg cannot compare versions across different schemes.

If a declared `version>=` constraint uses a different version scheme than the one used in the baseline version, vcpkg is unable to determine which version is greater or equal than the other.

For example, if you specify:

```json
{
  "dependencies": [
    {
      "name": "boost-regex",
      "version>=": "1.75.0"
    }
  ]
}
```

> vcpkg will emit the following error:

Reviewer: ```suggestion
vcpkg outputs the following error message:
```

   - Update the `version>=` constraint in your `vcpkg.json` to a version that uses a compatible scheme.
2. Override to the desired version:
   - If you need a specific version of boost-regex that uses a different versioning scheme, you can override it in your manifest.
   - Modify your `vcpkg.json` to include an overrides section specifying the desired version:
    ```json
    {
      "dependencies": [
        { "name": "boost-regex" }
      ],
      "overrides": [
        { "name": "boost-regex", "version": "1.75.0" }
      ]
    }
    ```

## <a name="shallow-clone-version-constraint"></a> Cause: Inadequate commit history in shallow clone
<!--
Steps to reproduce:
1. Perform a shallow clone of the vcpkg repository using Git with limited depth, for example: `git clone --depth=1 https://github.com/microsoft/vcpkg.git`.
2. In your project, create or edit the `vcpkg.json` file to include a dependency with a specific version constraint.
3. from the project directory, vcpkg x-update-baseline --add-initial-baseline
4. Run `vcpkg install`
-->

> When vcpkg is cloned with a limited commit history (shallow clone), it lacks the necessary commit history to resolve specific version constraints or baselines. This results in errors where vcpkg cannot find the required versions of packages or the specified baseline, as the commit history needed for these versions or baselines is not available in the shallow clone.

Reviewer: ```suggestion
When vcpkg is cloned with a limited commit history (shallow clone), it lacks the necessary commit history to resolve specific version constraints or baselines. The Git tree-object hashes used by vcpkg to retrieve specific versions of ports are only available when the full commit history is checked out. vcpkg detects when it has been cloned into a shallow repository and produces an error message when it fails to retrieve a port version.
```
